File: 1300/476B/main.cpp
#include <bits/stdc++.h>
using namespace std;

long long nCr(int n, int r) 
{
    if (r < 0 || r > n) 
        return 0;

    r = min(r, n - r);
    long long res = 1;

    for (int i = 1; i <= r; i++)
        res = res * (n - r + i) / i;

    return res;
}

int main() 
{
    string a, b;
    cin >> a >> b;

    int sum1 = 0, sum2 = 0, k = 0;

    for (int i = 0; i < a.size(); i++) 
    {
        if (a[i] == '+') 
            sum1++;
        else 
            sum1--;

        if (b[i] == '?') 
            k++;
        else if (b[i] == '+') 
            sum2++;
        else 
            sum2--;
    }

    int needed = sum1 - sum2;

    if ((needed + k) % 2 != 0) 
    {
        cout << fixed << setprecision(12) << 0.0 << endl;
        return 0;
    }

    int x = (needed + k) / 2;

    if (x < 0 || x > k) 
    {
        cout << fixed << setprecision(12) << 0.0 << endl;
        return 0;
    }

    long long fav = nCr(k, x);
    double total = pow(2, k);

    cout << fixed << setprecision(12) << fav / total << endl;

    return 0;
}

File: 1300/476B/main.java
import java.util.*;
public class main
{
    public static long nCr(int n, int r) 
    {
        if (r < 0 || r > n) 
            return 0;
            
        r = Math.min(r, n - r);
        long res = 1;
        for (int i = 1; i <= r; i++)
            res = res * (n - r + i) / i;
        
        return res;
    }
    
    public static void main(String args[])
    {
        Scanner sc = new Scanner(System.in);
        
        String a = sc.next();
        String b = sc.next();
        
        int sum1 = 0;
        int sum2 = 0;
        int k = 0;
        
        for(int i=0;i<a.length();i++)
        {
            if(a.charAt(i) == '+')
                sum1++;
            else
                sum1--;
            
            if(b.charAt(i) == '?')
                k++;
            else if(b.charAt(i) == '+')
                sum2++;
            else
                sum2--;
        }
        
        int needed = sum1 - sum2;
        
        if((needed+k)%2 != 0)
        {
            System.out.println(0.0);
            return ;
        }
        
        int x = (needed+k)/2;
        
        if(x < 0 || x > k)
        {
            System.out.println(0.0);
            return ;
        }
        
        long fav = nCr(k, x);
        double total = Math.pow(2, k);
        
        System.out.println(fav/total);
    }
}